pkg/phlaredb/block_querier.go

```diff
@@ -905,6 +938,90 @@ func (b *singleBlockQuerier) SelectMatchingProfiles(ctx context.Context, params
 	return iter.NewMergeIterator(maxBlockProfile, false, iters...), nil
 }
 
+// Series selects the series labels from this block.
+//
+// Note: It will select ALL the labels in the block, not necessarily just the
+// subset in the time range SeriesRequest.Start to SeriesRequest.End.
+func (b *singleBlockQuerier) Series(ctx context.Context, params *ingestv1.SeriesRequest) ([]*typesv1.Labels, error) {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "Series Block")
+	defer sp.Finish()
+
+	err := b.Open(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	selectors, err := parseSelectors(params.Matchers)
+	if err != nil {
+		return nil, err
+	}
+
+	names, err := b.index.LabelNames()
+	if err != nil {
+		return nil, err
+	}
+
+	labelNamesFilter := make(map[string]struct{}, len(params.LabelNames))
+	for _, n := range params.LabelNames {
+		labelNamesFilter[n] = struct{}{}
+	}
+	names = lo.Filter(names, func(name string, _ int) bool {
+		_, ok := labelNamesFilter[name]
+		return ok
+	})
+
+	var postingsIters []iter.SeekIterator[storage.SeriesRef, storage.SeriesRef]
```

Comment:
I would iterate through the postings right away, rather than keep the iterators open and do it later.